### PR TITLE
ARROW-9266: [Python][Packaging] Enable S3 support in macOS wheels

### PR DESCRIPTION
**Pull request by kszucs**
Depends on #8304

### CODE REVIEW

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-osx-mavericks-cp36m wheel-osx-high-sierra-cp36m

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-osx-*

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-osx-*

**Comment by nealrichardson** [comment]
@kszucs please rebase

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-osx-*

File: dev/tasks/python-wheels/osx-build.sh [hunk 1]
```diff
@@ -67,6 +67,7 @@ function build_wheel {
           -DARROW_GANDIVA=OFF \
           -DARROW_GRPC_USE_SHARED=OFF \
           -DARROW_HDFS=ON \
+          -DARROW_S3=${ARROW_S3} \
```
Comment (kou) [inline dev/tasks/python-wheels/osx-build.sh:70]: Could you keep alphabetical order?

File: dev/tasks/python-wheels/osx-build.sh [hunk 2]
```diff
@@ -103,6 +104,7 @@ function build_wheel {
     export PYARROW_WITH_DATASET=1
     export PYARROW_WITH_FLIGHT=1
     export PYARROW_WITH_HDFS=1
+    export PYARROW_WITH_S3=${ARROW_S3}
```
Comment (kou) [inline dev/tasks/python-wheels/osx-build.sh:107]: Ditto.

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-osx-*

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-osx-*

**Comment by kszucs** [comment]
The successful builds take between ~42-45 minutes but occasionally it exceeds the 50m build time limit. Sadly we can't utilize travis' build cache since crossbow always creates new branches and the fallback default branch won't have any cache entries.

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-osx-mavericks-cp38

**Comment by corleyma** [comment]
> The successful builds take between ~42-45 minutes but occasionally it exceeds the 50m build time limit. Sadly we can't utilize travis' build cache since crossbow always creates new branches and the fallback default branch won't have any cache entries.

Does this explain why some of the nightly builds seem to be intermittently missing certain macOS wheels?  E.g., sometimes for example there's no cp38 wheel, only for that wheel to come back with the next nightly publish.

**Comment by nealrichardson** [comment]
> Sadly we can't utilize travis' build cache since crossbow always creates new branches and the fallback default branch won't have any cache entries.

We *could* run something on `master` on the crossbow repo to warm the cache.

**Comment by corleyma** [comment]
I think @nealrichardson is right: my understanding is that travis caching will try to pull from default branch if no cache is available for current branch, so the warming strategy could work.

**Comment by kszucs** [comment]
> > Sadly we can't utilize travis' build cache since crossbow always creates new branches and the fallback default branch won't have any cache entries.
> 
> We _could_ run something on `master` on the crossbow repo to warm the cache.

We would need to merge the latests tasks (multiple .travis.yml files) generated by crossbow to master automatically which would be rather complicated. Crossbow also creates diverged branches from master at the moment.

**Comment by kszucs** [comment]
> Does this explain why some of the nightly builds seem to be intermittently missing certain macOS wheels? E.g., sometimes for example there's no cp38 wheel, only for that wheel to come back with the next nightly publish.

Could you send me build links? I don't entirely see your question.

**Review by kszucs — COMMENTED** [review]
+1, merging.

**Pull request closed without merging**